homeassistant/components/system_bridge/media_source.py

```diff
@@ -0,0 +1,217 @@
+"""System Bridge Media Source Implementation."""
+from __future__ import annotations
+
+import asyncio
+
+import async_timeout
+from systembridgeconnector.models.media_directories import MediaDirectories
+from systembridgeconnector.models.media_files import File as MediaFile
+
+from homeassistant.components.media_player.const import MEDIA_CLASS_DIRECTORY
+from homeassistant.components.media_source.const import (
+    MEDIA_CLASS_MAP,
+    MEDIA_MIME_TYPES,
+)
+from homeassistant.components.media_source.models import (
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+    PlayMedia,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_KEY, CONF_HOST, CONF_PORT
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+from .coordinator import SystemBridgeDataUpdateCoordinator
+
+
+async def async_get_media_source(hass: HomeAssistant) -> MediaSource:
+    """Set up SystemBridge media source."""
+    return SystemBridgeSource(hass)
+
+
+class SystemBridgeSource(MediaSource):
+    """Provide System Bridge media files as a media source."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+    ) -> None:
+        """Initialize source."""
+        super().__init__(DOMAIN)
+        self.name = "System Bridge"
+        self.hass: HomeAssistant = hass
+
+    async def async_resolve_media(
+        self,
+        item: MediaSourceItem,
+    ) -> PlayMedia:
+        """Resolve media to a url."""
+        entry_id, path, mime_type = item.identifier.split("~~", 2)
+        entry = self.hass.config_entries.async_get_entry(entry_id)
+        if entry is None:
+            raise ValueError("Invalid entry")
+        path_split = path.split("/", 1)
+        return PlayMedia(
+            f"{_build_base_url(entry)}&base={path_split[0]}&path={path_split[1]}",
+            mime_type,
+        )
+
+    async def async_browse_media(
+        self,
+        item: MediaSourceItem,
+    ) -> BrowseMediaSource:
+        """Return media."""
+        if not item.identifier:
+            return self._build_bridges()
+
+        if "~~" not in item.identifier:
+            entry = self.hass.config_entries.async_get_entry(item.identifier)
+            if entry is None:
+                raise ValueError("Invalid entry")
+            coordinator: SystemBridgeDataUpdateCoordinator = self.hass.data[DOMAIN].get(
+                entry.entry_id
+            )
+            await coordinator.async_get_media_directories()
+
+            async with async_timeout.timeout(20):
+                while coordinator.data.media_directories is None:
+                    await asyncio.sleep(1)
```

Comment:
Why do it like this? Can't you just return data? Otherwise use an event to notify when data is available ?

Reply:
The websocket asyncronously receives the event which is passed to the coordinator. This waits until this attribute is set from the coordinator event handler.

I'm not sure how I would return a `BrowseMediaSource` without pausing the function here. Do you know how I would use an event here? Any examples?

Reply:
All messages send over a websocket connection are asynchronous. That means you need to implement an RPC mechanism where you match requests with responses. 

Add an identifier to outgoing messages and create a result future to the server [like this](https://github.com/home-assistant-libs/python-matter-server/blob/main/matter_server/client/client.py#L113-L124)

When a message comes in and it's a result message, pass it to the result future [like this](https://github.com/home-assistant-libs/python-matter-server/blob/main/matter_server/client/client.py#L320-L342)

Reply:
Done 😺